### ISSUE TITLE
COPR repo for EL7

### ISSUE THREAD

**Issue by scx** [Beep6581/RawTherapee#5115]
Some time ago I created COPR repo for EL7:
https://copr.fedorainfracloud.org/coprs/scx/rawtherapee/
Today, I updated the rawtherapee package to the latest version.

This significantly simplifies the installation of RawTherapee on RHEL 7 / CentOS 7 / Scientific Linux 7 / Oracle Linux 7.

See also: Flatpak package - https://github.com/Beep6581/RawTherapee/issues/5113

**Comment by scx** [comment]
### RPM SPECs

`libtiff4-opt.spec`:
[libtiff4-opt.spec.txt](https://github.com/Beep6581/RawTherapee/files/2708465/libtiff4-opt.spec.txt)

`rawtherapee.spec`:
[rawtherapee.spec.txt](https://github.com/Beep6581/RawTherapee/files/2708466/rawtherapee.spec.txt)

**Comment by Beep6581** [comment]
Hey @scx 

Why would someone use this COPR repo instead of the official repo which already handles RawTherapee builds?
https://rawtherapee.com/downloads

> This significantly simplifies the installation of RawTherapee on RHEL 7 / CentOS 7 / Scientific Linux 7 / Oracle Linux 7.

How so, what are the steps?

**Comment by scx** [comment]
@Beep6581 

> Why would someone use this COPR repo instead of the official repo which already handles RawTherapee builds?

From what I see, for Linux you only provide AppImage packages. However, a lot of people prefer native packages/repositories (RPM/Yum, RPM/Zypper, DEB/Apt) instead of universal ones (Flatpak/Flathub, Snap/Snapcraft, AppImage/AppImageHub).
There are already native packages for [many distributions](https://pkgs.org/download/rawtherapee), but this is not the case with EL7 due to the requirement of `libtiff` >= _4.0.4_.
With the help of [COPR](https://copr.fedorainfracloud.org) and [OBS](https://openbuildservice.org), you could easily provide native packages for the most significant distributions, such as EL, Fedora, SLE, openSUSE, Mageia, Ubuntu, Debian, etc.
Going this way, you can easily provide continuous updates for all major distributions. In the case of AppImage, this is a little bit more [complicated](https://discourse.appimage.org/t/how-to-update-appimage/225/2) and there is no guarantee that the new version of software will start on the [given distribution](https://github.com/Beep6581/RawTherapee/issues/4687).

> How so, what are the steps?

The installation instruction is provided on the [linked page](https://copr.fedorainfracloud.org/coprs/scx/rawtherapee).
Currently it is:
```
cd '/etc/yum.repos.d/'

curl 'https://copr.fedorainfracloud.org/coprs/scx/libtiff4-opt/repo/epel-7/scx-libtiff4-opt-epel-7.repo' > 'scx-libtiff4-opt-epel-7.repo'

curl 'https://copr.fedorainfracloud.org/coprs/scx/rawtherapee/repo/epel-7/scx-rawtherapee-epel-7.repo' > 'scx-rawtherapee-epel-7.repo'

yum install 'rawtherapee'
```

It is much simpler and more elegant way than [the official one](http://rawpedia.rawtherapee.com/Linux#CentOS).

### Rebuilding

Rebuilding is pretty easy:

#### rpmbuild

```
rpmbuild --rebuild rawtherapee-5.5-1.1.el7.src.rpm
```

or:

```
rpm -ivh rawtherapee-5.5-1.1.el7.src.rpm
```

```
rpmbuild -ba ~/rpmbuild/SPECS/rawtherapee.spec
```

#### mock

```
mock -r epel-7-x86_64 --rebuild rawtherapee-5.5-1.1.el7.src.rpm
```

**Comment by scx** [comment]
TL;DR

Although there are [semi-official repositories](http://download.opensuse.org/repositories/home:/rawtherapee/) for Fedora, this is not the case with EL7 due to the requirement of `libtiff` >= _4.0.4_.
I'm trying to fill this gap with my [COPR repository](https://copr.fedorainfracloud.org/coprs/scx/rawtherapee).

**Comment by Beep6581** [comment]
Thank you for your contribution. I have a few questions.

> From what I see, for Linux you only provide AppImage packages.

We do provide native packages through native repositories using OBS.

> It is much simpler and more elegant way than the official one.

You're comparing instructions for self-compiling RawTherapee to instructions for downloading a pre-compiled RPM binary. Those are quite different things.

> Although there are semi-official repositories for Fedora

1. What makes them semi-official?
1. What could we or the maintainer do to make them fully official?

> Although there are semi-official repositories for Fedora, this is not the case with EL7 due to the requirement of libtiff >= 4.0.4. I'm trying to fill this gap with my COPR repository.

Agreed. I will add your RHEL7 repository. 

1. How does a user get a debug build type?
1. Which version of GTK+ is used?

**Comment by scx** [comment]
@Beep6581 

> We do provide native packages through native repositories using OBS.

From what I understand, you personally provide only the AppImage package, while others are responsible for other packages/repositories (e.g. Marcin Bajor for Debian packages, Dariusz Duma for Ubuntu packages, etc.). 

> You're comparing instructions for self-compiling RawTherapee to instructions for downloading a pre-compiled RPM binary. Those are quite different things.

You're right, but at the moment there are no RPM packages for EL7 (maybe except for the very outdated ones that stuck on version 4.2), so you have to build them on your own.

> What makes them semi-official?

By "semi-official" I meant that they are maintained by other people (Marcin Bajor, Dariusz Duma, etc.). However, this may have a different meaning. The official packages are those that are in the distribution repository.

> What could we or the maintainer do to make them fully official?

The rawtherapee package is already in the Fedora repository.
https://koji.fedoraproject.org/koji/packageinfo?packageID=10105
https://src.fedoraproject.org/rpms/rawtherapee

In the case of EL7 it would also be possible, but unfortunately we have a problematic dependence of the `libtiff` >= *4.0.4*.
You could provide a RPM package with bundled `libtiff` library (`libtiff.so.5`) by installing it in `%{_libdir}/rawtherapee/` (e.g. `/usr/lib64/rawtherapee/` for x86_64), then try to become a Fedora [contributor](https://fedoraproject.org/wiki/How_to_be_a_successful_contributor). However, I don't know if this is worth it.

Things you need to pay attention to:

1. Devtoolset:
```
BuildRequires:  devtoolset-7-gcc
BuildRequires:  devtoolset-7-gcc-c++
BuildRequires:  devtoolset-7-toolchain
BuildRequires:  devtoolset-7-libatomic-devel
```

```
. /opt/rh/devtoolset-%{devtoolset}/enable
```

2. CMake options for `libtiff`:
```
        -DTIFF_INCLUDE_DIR="${LIBTIFF4_INCLUDE_PATH}" \
        -DTIFF_LIBRARIES="${LIBTIFF4_LIBRARY_PATH}" \
        -DTIFF_LIBRARY="${LIBTIFF4_LIBRARY_PATH}/libtiff.so" \
```

3. RPATH:
```
        -DCMAKE_INSTALL_RPATH="${LIBTIFF4_LIBRARY_PATH}" \
```

4. Filtering automatic dependencies:
```
%filter_from_requires /^libtiff\.so\.5.*$/d; /^libtiffxx\.so\.5.*$/d
%filter_provides_in %{_libdir}/%{name}
%filter_setup
```

> How does a user get a debug build type?

As this is based on the official Fedora package, you can just install `rawtherapee-debuginfo` to get debug symbols.

> Which version of GTK+ is used?

Gtk+3. Please keep in mind that GNOME 3 is the official desktop environment for RHEL 7.

**Comment by Beep6581** [comment]
> By "semi-official" I meant that they are maintained by other people

Those "other people" work closely with us and have our full blessing and support :)

> The official packages are those that are in the distribution repository.

I hope the latest release is available in all official distro repositories. It is up to our users to ensure that they file a request in their distro's issue tracker if the latest release is missing from their package manager. My interest goes as far as ensuring that there are clear, simple to follow and unambiguous download instructions available from our Downloads page to get the latest release.

> Gtk+3.

Which version of GTK3?

**Comment by scx** [comment]
> Which version of GTK3?

Currently 3.22.30, but it may be rebased to >= 3.24.1 in the future.
https://git.centos.org/summary/rpms!gtk3.git

**Comment by Beep6581** [comment]
@scx Copr repo added, please verify:
https://rawtherapee.com/downloads

**Comment by scx** [comment]
@Beep6581 

> @scx Copr repo added, please verify:
> https://rawtherapee.com/downloads

It looks fine. Thanks a lot!

**Comment by Beep6581** [comment]
Thank you.

**Comment by scx** [comment]
I've just created RPM packages for **RawTherapee** _5.6_ for EL7 (Red Hat Enterprise Linux 7, CentOS 7, Scientific Linux 7, Oracle Linux 7).

Sorry for the delay, but there were some obstacles along the way.

## 1. CMake

According to the [CMakeLists.txt](https://github.com/Beep6581/RawTherapee/blob/5.6/CMakeLists.txt#L5), **RawTherapee** requires `cmake` >= _2.8.8_.

However, I was unable to build **RawTherapee** using `cmake` _2.8.12.2_.
```]
CMake Error at rtdata/CMakeLists.txt:52 (install):
  install FILES given directory
  "/home/scx/rpmbuild/BUILD/rawtherapee-5.6/rtdata/images/png" to install.
```

Fortunately, it works with `cmake` _3.13.4_.

## 2. AppData

The current AppData file is simply not valid. It is because of [wrong version scheme](https://github.com/Beep6581/RawTherapee/blob/5.6/com.rawtherapee.RawTherapee.appdata.xml#L39-L40).
```
$ appstream-util validate-relax --nonet com.rawtherapee.RawTherapee.appdata.xml 
com.rawtherapee.RawTherapee.appdata.xml: FAILED:
• tag-invalid           : <release> versions are not in order [5.6 before 5.6-rc2]
Validation of files failed
```

First of all, you should know that AppSteam requires to put releases in order. It doesn't define which version number is valid or not, but it uses literally the same [version compare algorithm](https://github.com/ximion/appstream/blob/12c1038bb0b55535f84e9900de1968619ac430cb/src/as-utils.c#L981-L1103) that is used by [RPM](https://github.com/rpm-software-management/rpm/blob/effa0b02a636a789f5d00f586a18dd961104623e/lib/rpmvercmp.c#L12-L137).
Moreover, this is explicitly stated in [documentation](https://www.freedesktop.org/software/appstream/docs/api/appstream-as-utils.html#as-utils-compare-versions).

That is why the best way is to use [Fedora versioning policy](https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/), especially the part about [prereleases](https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_versioning_prereleases_with_tilde).
In the other words, you should use tilde (`~`) instead of hyphen-minus (`-`).
```
$ rpmdev-vercmp '5.5' '5.6'
5.5 < 5.6
$ rpmdev-vercmp '5.6' '5.6'
5.6 == 5.6
$ rpmdev-vercmp '5.6-rc2' '5.6'
5.6-rc2 > 5.6
$ rpmdev-vercmp '5.6~rc2' '5.6'
5.6~rc2 < 5.6
$ rpmdev-vercmp '5.6~rc1' '5.6~rc2'
5.6~rc1 < 5.6~rc2
```

I prepared a temporary patch:
```
diff -Naur RawTherapee-5.6.orig/com.rawtherapee.RawTherapee.appdata.xml RawTherapee-5.6/com.rawtherapee.RawTherapee.appdata.xml
--- RawTherapee-5.6.orig/com.rawtherapee.RawTherapee.appdata.xml	2019-04-20 21:56:55.000000000 +0200
+++ RawTherapee-5.6/com.rawtherapee.RawTherapee.appdata.xml	2019-04-23 17:07:10.733956384 +0200
@@ -36,8 +36,8 @@
     <launchable type="desktop-id">rawtherapee.desktop</launchable>
     <releases>
         <release version="5.6" date="2019-04-20" type="stable"></release>
-        <release version="5.6-rc2" date="2019-04-17" type="development"></release>
-        <release version="5.6-rc1" date="2019-04-10" type="development"></release>
+        <release version="5.6~rc2" date="2019-04-17" type="development"></release>
+        <release version="5.6~rc1" date="2019-04-10" type="development"></release>
         <release version="5.5" date="2018-12-17" type="stable"></release>
     </releases>
     <provides>
```

Unfortunately, the LTS distributions can still have a problem with this file, because of a bug in the old version of `libappstream-glib`. There is nothing we can do upstream about it. However, it can be easily fixed at build time:
```
xmlstarlet ed --inplace -d '/component/releases/release[position()>1][contains(@version,"~")]' "com.rawtherapee.RawTherapee.appdata.xml"
```

This is exactly what I have done in my [RPMs](https://copr.fedorainfracloud.org/coprs/scx/rawtherapee/build/886982/).

**Comment by scx** [comment]
I've updated the RPM packages for RawTherapee to version 5.7.
CentOS 7 and Fedora 29-31 + Rawhide are currently supported.
https://copr.fedorainfracloud.org/coprs/scx/rawtherapee/
https://copr.fedorainfracloud.org/coprs/scx/rawtherapee/build/1040388/

These packages contain upstreamed patches for external editor support and XDG-related stuff.

**Comment by scx** [comment]
I've provided packages for EL8 (aarch64, ppc64le, x86_64):
https://copr.fedorainfracloud.org/coprs/scx/rawtherapee/build/1062155/
